pkg/controllers/execution/execution_controller.go

```diff
@@ -185,6 +188,13 @@ func (c *Controller) syncToClusters(clusterName string, work *workv1alpha1.Work)
 	for _, manifest := range work.Spec.Workload.Manifests {
 		workload := &unstructured.Unstructured{}
 		err := workload.UnmarshalJSON(manifest.Raw)
+		reference, _ := ref.GetReference(c.Client.Scheme(), workload)
```

Comment:
I suppose we can add a helper function at such as `pkg/util/helper/work.go`, or just to the controller, like:
```golang
// GenEventRefObj sets the UID(.metadata.uid) that might be missing for fire events.
// Do nothing if the UID already exist, otherwise set the UID from annotation.
func GenEventRefObj(resource *unstructured.Unstructured) *unstructured.Unstructured {
	if len(resource.GetUID()) > 0 {
		return resource
	}

	// set UID according to ResourceTemplateUIDAnnotation.
	// TODO

	return resource
}
```
or 
```golang
// GenEventRef returns the event reference.
func GenEventRef(resource *unstructured.Unstructured) *corev1.ObjectReference {

}
```

It will be good to re-use this function if we need to fire an event in another place.
And we can add a unit test for it too.

Reply:
I prefer `GenEventRef `, `recorder` converts the object to an ObjectReference and no longer uses the object later. `GenEventRef` is enough and Avoid deep copy of the object。